crates/next-core/src/next_root_params/mod.rs

```diff
@@ -0,0 +1,244 @@
+                            // directory. TODO: make sure this error is
+                            // consistent with webpack
+                            Self::invalid_import_map_result(
+                                "'next/root-params' can only be used inside the App Directory."
+                                    .into(),
+                            )
+                        }
+                        _ => {
+                            // In general, the compiler should prevent importing 'next/root-params'
+                            // from client modules, but it doesn't catch
+                            // everything. If an import slips through
+                            // our validation, make it error.
+                            Self::invalid_import_map_result(
+                                "'next/root-params' cannot be imported from a Client Component \
+                                 module. It should only be used from a Server Component."
+                                    .into(),
+                            )
+                        }
+                    },
+                    Either::Right(_) => {
+                        // In general, the compiler should prevent importing 'next/root-params' from
+                        // client modules, but it doesn't catch everything. If an
+                        // import slips through our validation, make it error.
+                        Self::invalid_import_map_result(
+                            "'next/root-params' cannot be imported from a Client Component \
+                             module. It should only be used from a Server Component."
+                                .into(),
+                        )
+                    }
+                }
+            }
+        })
+    }
+
+    #[turbo_tasks::function]
+    async fn valid_import_map_result(
+        collected_root_params: ResolvedVc<CollectedRootParams>,
+    ) -> Result<Vc<ImportMapResult>> {
+        let collected_root_params = collected_root_params.await?;
+
+        // Generate a virtual 'next/root-params' module based on the root params we collected.
+        let module_content =
+            // If there's no root params, export nothing.
+            if collected_root_params.is_empty() {
+                "export {}".to_string()
+            } else {
+                iter::once(formatdoc!(
+                    r#"
+                        import {{ getRootParam }} from 'next/dist/server/request/root-params';
+                    "#,
+                ))
+                .chain(collected_root_params.iter().map(|param_name| {
+                    formatdoc!(
+                        r#"
+                            export function {PARAM_NAME}() {{
+                                return getRootParam('{PARAM_NAME}');
+                            }}
+                        "#,
+                        PARAM_NAME = param_name,
```

Comment:
```suggestion
                        PARAM_NAME = StringifyJs(param_name),
```

It needs different arguments for the identifier and the string argument

Reply:
i'm gonna pull these out into a follow-up, this also needs a bunch of tests and logic to generate a relatively human friendly function name for the getter, and using a param that isn't a valid identifier is uncommon